From 9d53b39ceb549bd8efbe6eacf1964c6d282d619c Mon Sep 17 00:00:00 2001
From: Rama Chavali <rama.rao@salesforce.com>
Date: Fri, 3 Dec 2021 11:07:47 +0530
Subject: revert fault filter change (#36300)

* remove fault and cors filters

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* add cors filter bacl

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* fix tests

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>

* revert fault filter change

Signed-off-by: Rama Chavali <rama.rao@salesforce.com>
---
 pilot/pkg/networking/core/v1alpha3/listener.go      | 9 ++++-----
 pilot/pkg/networking/core/v1alpha3/listener_test.go | 2 +-
 2 files changed, 5 insertions(+), 6 deletions(-)

diff --git a/pilot/pkg/networking/core/v1alpha3/listener.go b/pilot/pkg/networking/core/v1alpha3/listener.go
index cd28ea17f3..20ad0f8e87 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener.go
@@ -1302,10 +1302,8 @@ func buildHTTPConnectionManager(listenerOpts buildListenerOpts, httpOpts *httpLi
 
 	routerFilterCtx := configureTracing(listenerOpts, connectionManager)
 
-	// Add fault filter at the top.
-	filters := make([]*hcm.HttpFilter, len(httpFilters)+1)
-	filters[0] = xdsfilters.Fault
-	copy(filters[1:], httpFilters)
+	filters := make([]*hcm.HttpFilter, len(httpFilters))
+	copy(filters, httpFilters)
 
 	if features.MetadataExchange {
 		filters = append(filters, xdsfilters.HTTPMx)
@@ -1324,7 +1322,8 @@ func buildHTTPConnectionManager(listenerOpts buildListenerOpts, httpOpts *httpLi
 		filters = append(filters, xdsfilters.Alpn)
 	}
 
-	filters = append(filters, xdsfilters.Cors)
+	// TypedPerFilterConfig in route needs these filters.
+	filters = append(filters, xdsfilters.Fault, xdsfilters.Cors)
 	filters = append(filters, listenerOpts.push.Telemetry.HTTPFilters(listenerOpts.proxy, listenerOpts.class)...)
 	filters = append(filters, xdsfilters.BuildRouterFilter(routerFilterCtx))
 
diff --git a/pilot/pkg/networking/core/v1alpha3/listener_test.go b/pilot/pkg/networking/core/v1alpha3/listener_test.go
index e1d41bd8ee..6e43058145 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener_test.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener_test.go
@@ -2261,7 +2261,7 @@ func verifyInboundEnvoyListenerNumber(t *testing.T, l *listener.Listener) {
 			t.Fatalf("expected HTTP filter, found none")
 		}
 
-		expect := []string{xdsfilters.Fault.Name, xdsfilters.MxFilterName, xdsfilters.Cors.Name, xdsfilters.Router.Name}
+		expect := []string{xdsfilters.MxFilterName, xdsfilters.Fault.Name, xdsfilters.Cors.Name, xdsfilters.Router.Name}
 		got := getHCMFilters(t, f)
 		if !reflect.DeepEqual(expect, got) {
 			t.Fatalf("expected http filters %v, found %v", expect, got)
-- 
2.35.3

